### PR TITLE
Adding patch embeddings

### PR DESCRIPTION
**Pull request by stergioc**
Related to #19 

I added a patch embedding functionality for WSIs. Specifically, patches are retrieved with a specific `patch_width` from a specific `magnification` level. Then they are pushed as batches of `batch_size` into one of three embedding models. The resulting embeddings are stored as an additional image within the spatialdata object with index the class of the embedding model. 

in order to test it you can run the following:

```python
import spatialdata_plot
from sopa.io import wsi
from sopa.segmentation.tissue import hsv_otsu
from sopa.embedding.patches import embed_wsi_patches

slide = wsi('https://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/CMU-1.svs')
hsv_otsu(slide, "CMU-1")
embed_wsi_patches(slide, "CMU-1", "Resnet50Features", 10, 224, 64, num_workers=8)

slide.pl.render_images("Resnet50Features", channel=[1, 2, 3]).pl.show()
```

the output of the plotting function using only the first 3 channels of ResNet features looks like this:

### CODE REVIEW

**Comment by quentinblampey** [comment]
Hello @stergioc, thanks for the PR!
I tried using a little bit more the `Patches2D` class, which has some convenient functionnalities. I also added an assert to check that the model we ask for is existing

I have a few comments, see below:

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:42]: Here `image_key` is not accessible, so we can't log it

Comment (quentinblampey) [inline sopa/embedding/patches.py:42]: Here `image_key` is not accessible, so we can't log it

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:108]: Unused variable. Do you want to do something with it or we delete it?

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:95]: Unused argument, do we need patch overlap?

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:18]: Can you add some typing? Same for the below function

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:126]: Here "pixels" is hardcoded. Instead, we should use the intrinsic coordinate system of the image. I'll add it myself later

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:58]: `"image"` is hardcoded, we have to use `next(iter(patch[f"scale{level}"]))` to get the first (and unique) DataArray of the Dataset

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:28]: Can you add a small docstring?

File: sopa/embedding/models/__init__.py
Comment (quentinblampey) [inline sopa/embedding/models/__init__.py:2]: For consistency, I would rename the module `histo_ssl`, and the class `HistoSSLFeatures`, what do you think?

**Review by quentinblampey — CHANGES_REQUESTED** [review]
I added some comments :)

File: sopa/embedding/patches.py
Comment (stergioc) [reply sopa/embedding/patches.py:42]: What if I put this log after the call of the function and check if it returns false, instead of putting it here?

Comment (quentinblampey) [reply sopa/embedding/patches.py:53]: Yes, sounds good! Indeed it will avoid passing the argument to this function

File: sopa/embedding/patches.py
Comment (quentinblampey) [inline sopa/embedding/patches.py:44]: What is the purpose of `mppdiff`? It seems to be also unused

Comment (stergioc) [reply sopa/embedding/patches.py:44]: Hmm that is a good question. Something looks weird here. I will take a closer look.

Comment (quentinblampey) [reply sopa/embedding/patches.py:44]: Ok thanks!

**Comment by quentinblampey** [comment]
Just some minor changes needed but it already looks very good!

**Comment by quentinblampey** [comment]
One minor question: what is the meaning of `obj`? "Objective"?

**Comment by quentinblampey** [comment]
It's looking better, but we still have a slight shift in the plot. I'll investigate on this later
image

**Review by quentinblampey — APPROVED** [review]

**Comment by quentinblampey** [comment]
Okay it's better now @stergioc. We still have a small shift (see below), but this is only a plotting issue. Indeed, by default matplotlib show a pixel centered (from -0.5 to 0.5), so there is a shift of 0.5 pixels. Usually, this is not visible, but here I have a transformation of `Scale(1000, 1000)`, so it makes a shift of 500 pixels during the plot. I opened an issue in spatialdata-plot here: https://github.com/scverse/spatialdata-plot/issues/216

image

File: sopa/embedding/patches.py
Comment (stergioc) [inline sopa/embedding/patches.py:179]: Just a quick comment since I see you are changing things. This info message should write something like, WSI embeddings saved in ...

**Comment by quentinblampey** [comment]
Of course the shift is less visible with a smaller `patch_width`. I just did some final clean up (docstrings, typing, ...). Ready for merging when I have your approval @stergioc 

image

**Comment by stergioc** [comment]
Amazing @quentinblampey! I am running some tests on a bunch of in-house slides and testing the embedding speed. I will let you know when it is done :)

File: sopa/io/histopathology.py
Comment (quentinblampey) [inline sopa/io/histopathology.py:102]: Does it happen to have multiple extensions (like `.ome.tif`) or is it always one? Because `.stem` only removes the "last" extension, i.e. `image.ome.tif` would become `image.ome`

Comment (stergioc) [reply sopa/io/histopathology.py:102]: hmm I did this because there are some filename that include `.` before the file extension. Here is an example from TCGA:  `TCGA-73-4658-01A-01-BS1.2859e392-66d8-407e-85db-25bf4e4c5ff3.svs` but there are more examples like this at the in-house databases with more segments in the filename that are separated with a `.`. Maybe we need a filetype detection function potentially the detect_format function of tiffslide can help?

Comment (quentinblampey) [reply sopa/io/histopathology.py:102]: Maybe we could implement something like this, yes. Do you have a list of possible WSI extensions? Mostly, which one has a "double" extension?

But maybe we can keep your current implementation for now and open a new PR later if needed?

**Comment by stergioc** [comment]
**Note:** at the moment the current implementation of the  OpenSlideStore works only for `zarr==2.14.2`. Not sure atm what would require in order to bump it to the current latest version of `zarr==2.17.0`.

**Comment by stergioc** [comment]
I reverted the merge with wsi-openslide in order to first fix some of the issues that I identified before merging it with the wsi branch. We can stick with the tiffslide at the moment here.

**Comment by quentinblampey** [comment]
Looking good to me @stergioc! Indeed we can do a PR later for openslide. I can merge now if it's all good on your side!

**Comment by stergioc** [comment]
Go ahead yes it's all good from my side.

**Pull request merged**
